### PR TITLE
[Backport] Changes to support retrieval of operations from translog based on specified range

### PR DESCRIPTION
**Pull request by saikaranam-amazon**
[Backport] Changes to support retrieval of operations from translog based on specified range
Signed-off-by: Sai Kumar <karanas@amazon.com>

### Description
[Backport] Changes to support retrieval of operations from translog based on specified range
 
### Issues Resolved
N/A
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by tlfeng** [comment]
start gradle check

File: server/src/main/java/org/opensearch/index/engine/Engine.java [hunk 1]
```diff
@@ -742,6 +742,15 @@ boolean assertSearcherIsWarmedUp(String source, SearcherScope scope) {
     public abstract Translog.Snapshot newChangesSnapshot(String source, MapperService mapperService,
                                                          long fromSeqNo, long toSeqNo, boolean requiredFullRange) throws IOException;
 
+    /**
+     * Creates a new history snapshot from either Lucene/Translog for reading operations whose seqno in the requesting
+     * seqno range (both inclusive).
+     */
+    public Translog.Snapshot newChangesSnapshot(String source, HistorySource historySource, MapperService mapperService,
```
Comment (nknize) [inline server/src/main/java/org/opensearch/index/engine/Engine.java:749]: Is this useful for anything other than the retention lease pruning used by CCR?

File: server/src/main/java/org/opensearch/index/translog/Translog.java [hunk 2]
```diff
@@ -614,7 +615,11 @@ final Checkpoint getLastSyncedCheckpoint() {
 
     // for testing
     public Snapshot newSnapshot() throws IOException {
-        return newSnapshot(0, Long.MAX_VALUE);
+        return newSnapshot(0, Long.MAX_VALUE, false);
```
Comment (nknize) [inline server/src/main/java/org/opensearch/index/translog/Translog.java:618]: Use the helper method added.

```suggestion
        return this.newSnapshot(0, Long.MAX_VALUE);
```

**Review by nknize — CHANGES_REQUESTED** [review]
First quick review is looking good. Just one minor nit; and a overall question if this is something that would be useful outside the translog retention lease pruning.

File: server/src/main/java/org/opensearch/index/engine/Engine.java (hunk 1 above)
Comment (saikaranam-amazon) [reply server/src/main/java/org/opensearch/index/engine/Engine.java:749]: Yes, currently used by only CCR

**Comment by saikaranam-amazon** [comment]
> and a overall question if this is something that would be useful outside the translog retention lease pruning.

Yes, Main use-case that it is currently addressing is to surface the translog ops based on history source

**Comment by CEHENKLE** [comment]
start gradle check

**Comment by CEHENKLE** [comment]
start gradle check

**Review by nknize — APPROVED** [review]
LGTM

**Pull request merged**